### ISSUE TITLE
🐛 BUG: Wrangler dev reporting logged out, fixed by `npx wrangler whoami`

### ISSUE THREAD

**Issue by poacher2k** [cloudflare/workers-sdk#2978]
### Which Cloudflare product(s) does this pertain to?

Wrangler

### What version of `Wrangler` are you using?

2.13.0

### What operating system are you using?

Mac

### Describe the Bug

When it's been a while (~1 hr ? a day ?) since I've run `wrangler dev`, it fails with the following error:
![image](https://user-images.githubusercontent.com/2255960/229467653-dff76e55-93b8-47ce-9d5b-74d0f507f7b8.png)

If I then run `npx wrangler whoami`, it correctly gets user settings and shows that I'm logged in. Subsequently, running `wrangler dev` again now succeeds.

**Comment by poacher2k** [comment]
> Can you reproduce this in multiple projects, or is it specific to a single one?

I'm running 2 workers at the same time, and I'm not sure if it's specific to both of them or if it's just that one of them is consistently failing _first_. I'll try waiting a while and running the other one to see if it's specific to the one worker.

**Comment by petebacondarwin** [comment]
So you are running them using some kind of parallelisation script? I think this means that they are in non-interactive mode, which is why it is not asking you to login.

When you just run either of the Workers does independently does it work oK?

Do you have any environment variables set like CLOUDFLARE_API_TOKEN, CF_API_TOKEN, CLOUDFLARE_API_KEY or CF_API_KEY?

Do you have a global auth config file? E.g. `~/.wrangler/config/default.toml` or similar in your XDG App path?

**Comment by poacher2k** [comment]
> So you are running them using some kind of parallelisation script? I think this means that they are in non-interactive mode, which is why it is not asking you to login.  

Yes, I'm using [turbo](https://turbo.build/repo). It fails the first time (usually at the beginning of the day), and then works on consecutive runs after running eg. `npx wrangler whoami` or `npx wrangler login`. 

> When you just run either of the Workers does independently does it work oK?

I'll try when some time has passed.

> Do you have any environment variables set like CLOUDFLARE_API_TOKEN, CF_API_TOKEN, CLOUDFLARE_API_KEY or CF_API_KEY?

Yes, I set `CLOUDFLARE_TOKEN` - can this be messing things up? It does, as mentioned, work on consecutive runs though 🤔

> Do you have a global auth config file? E.g. `~/.wrangler/config/default.toml` or similar in your XDG App path?

Yes, `~/.wrangler/config/default.toml` exists.

**Comment by petebacondarwin** [comment]
Hmm. So each `wrangler dev` command should be picking up the CLOUDFLARE_API_TOKEN variable and using that first.
Are you sure it is valid?

I was wonder if, maybe, the first time you run, it uses the CLOUDFLARE_API_TOKEN and fails. Then when you do a normal login, it is caching the correct API token in the global auth config file?

Can you try running with verbose debugging on to see if that helps?  `WRANGLER_LOG=debug npx wrangler dev`? I.e. set this environment variable.

**Comment by poacher2k** [comment]
> Hmm. So each `wrangler dev` command should be picking up the CLOUDFLARE_API_TOKEN variable and using that first. Are you sure it is valid?

Sorry, I realize I wasn't clear - I set `CLOUDFLARE_TOKEN` for use in my non-worker application. I rely on `wrangler login` for actually running the workers with needed permissions. I don't set `CLOUDFLARE_API_TOKEN` manually anywhere.

> I was wonder if, maybe, the first time you run, it uses the CLOUDFLARE_API_TOKEN and fails. Then when you do a normal login, it is caching the correct API token in the global auth config file?

Makes sense, but shouldn't it pick up the global auth config file during the first run too?

> Can you try running with verbose debugging on to see if that helps provide more info? `WRANGLER_LOG=debug npx wrangler dev`? I.e. set this environment variable.

Will do!

**Comment by penalosa** [comment]
@poacher2k Just quickly following up—did you have a chance to run this with verbose logging?

**Comment by poacher2k** [comment]
@penalosa : Just a quick follow-up to say this still occurs with `^3.1.1`

**Comment by poacher2k** [comment]
> Hey! 👋 Apologies for the delayed response here. I'm unable to see what's going wrong here from the verbose logging. Any chance you could put together a minimal reproduction/share your code here?

Here you go: https://github.com/poacher2k/workers-sdk-issue-2978-repro

**Comment by penalosa** [comment]
Having investigated this a bit, this is due to a race condition with both Wrangler processes trying to use the same Oauth refresh token at the same time. 

As a workaround, you can run `npx wrangler whoami`, which will refresh your auth tokens.

**Comment by poacher2k** [comment]
> As a workaround, you can run `npx wrangler whoami`, which will refresh your auth tokens.

Sadly this (and even just the default behavior now) makes it extra annoying, as it now automatically opens a browser tab asking me if I want to give Wrangler access etc. Before it would just fail and I could re-run it immediately - now I have to close the browser tab, and manually exit the processes before I can re-run it.

**Comment by poacher2k** [comment]
To anyone else experiencing this, I've managed to work around it by adding `sleep N`, eg. `sleep 3 && wrangler dev` to my workers, incrementing `N` by a couple for each of them to have them start sequentially instead of in parallel - works like a charm!

**Comment by petebacondarwin** [comment]
Closing as we don't plan to provide a way to avoid the race condition in `wrangler dev` - instead run a command like `wrangler whoami` before to refresh the tokens.
